### PR TITLE
feat: add pre-commit framework

### PR DESCRIPTION
**Pull request by jbampton**
### What this PR does / why we need it:

From the pre-commit site -> "Git hook scripts are useful for identifying simple issues before submission to code review. We run our hooks on every commit to automatically point out issues in code such as missing semicolons, trailing whitespace, and debug statements. By pointing these issues out before code review, this allows a code reviewer to focus on the architecture of a change while not wasting time with trivial style nitpicks."

refs #3289 

https://pre-commit.com/

The next image shows the final pre-commit run for this pull request:

Screen Shot 2021-02-22 at 3 14 50 am

### Pre-submission checklist:

* [X] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by jbampton** [comment]
All the hooks are listed here -> https://pre-commit.com/hooks.html

So we can add even more checks.   There is even a check we could add to add licenses to files that don't have them.

Some of the checks are commented out and can be added in future.

I added a few more standard tests seen in the next image:

Screen Shot 2021-02-22 at 5 26 51 am

File: .pre-commit-config.yaml
Comment (jbampton) [inline .pre-commit-config.yaml:60]: I was not able to get the `luacheck` to pass for all files -> `pre-commit run --all-files`

So I have commented it out for now.  

I think it's failing on these warning messages:

Screen Shot 2021-02-22 at 5 28 11 am

**Comment by jbampton** [comment]
When running `pre-commit run --all-files` the main tests now look like this:

Screen Shot 2021-02-22 at 7 47 13 am

**Comment by tokers** [comment]
@jbampton Well done! We can also apply them to the dashboard and ingress controller projects after this PR merge.

File: .travis/apisix_cli_test/common.sh
Comment (spacewander) [inline .travis/apisix_cli_test/common.sh:1]: The `common.sh` is expected to be sourced but not executed.

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
The pre-commit framework is good.
But does it add any value as we already check them in the CI? There are many duplicate rules between this framework and the CI files. I don't want to increase the maintain burden and make amateur harder to do contribute to it.

**Comment by jbampton** [comment]
This feature is very useful since it runs on your local machine when you do the commit and it runs very fast and gives immediate feedback to you so you don't have to wait for the CI to run before you get feedback.

Also this pre-commit framework automatically fixes a lot of the problems in the codebase for example it runs `black` on Python files and it fixes `trailing whitespace`.  It also fixes `end of files` or makes sure each file has a single blank line at the end.  It also fixes mixed line endings converting to `lf`.  

There are many other checks and they are all useful and most of these tests we don't even have on the CI.  

Pre-commit automatically fixes code for you so it speeds up development. 

Most of the top GitHub projects are using pre-commit even Apache Airflow.

**Comment by spacewander** [comment]
We are not a Python project. Can it fix the markdown / yaml automatically?

**Comment by Yiyiyimu** [comment]
Since we are using pre-commit, do we need to remove those lint checkers since they are duplicates now. We could use pre-commit as the lint checkers in Actions

**Comment by jbampton** [comment]
It does say that `pre-commit` can be run on CI -> https://pre-commit.com/#usage-in-continuous-integration

**Comment by Yiyiyimu** [comment]
> It does say that `pre-commit` can be run on CI -> https://pre-commit.com/#usage-in-continuous-integration

Cool! I think it would be better if we could do that in this PR

**Comment by membphis** [comment]
need to resolve the Conflicting files

**Comment by moonming** [comment]
any update? @jbampton

**Comment by jbampton** [comment]
I rebased this twice after @membphis  asked me and no one took notice  and it never got merged.

Now it's marked as stale  and it's conflicted again.

Anyway @spacewander never gave his approval.

Honestly I am starting to lose interest since I have worked on most of the APISIX projects but I'm not on the team.

**Comment by jbampton** [comment]
Hey @moonming I got an update for you 👍 

Apache Airflow uses a best in class `pre-commit` framework where they even have checks that automatically insert missing Apache Licenses on `git commit`.  So when this happens the `license check` or `checks` fail as `red` and the commit fails and then you need to do another commit with these new additions.

If you look from around line 33 to line 143 in the Airflow `.pre-commit-config.yaml` file you will see all the license code for all the different file types. Some file types are grouped togethor as they have the same `comment` syntax.

https://github.com/apache/airflow/blob/566127308f283e2eff29e8a7fbfb01f17a1cd18a/.pre-commit-config.yaml#L33

**Comment by jbampton** [comment]
> Our project is not a Python project. Can it fix the markdown / yaml automatically?

Hey @spacewander as I mentioned before I have worked with `overcommit` and `pre-commit` but on some other projects I did also work with `husky`.  `husky` is huge in the JavaScript world as it's mainly JS.

https://typicode.github.io/husky/#/
https://www.npmjs.com/package/husky
https://github.com/typicode/husky

I think for sure `pre-commit` is the best and more research and work is needed to get all this working well.  And we need more docs on all of this.

Auto fixing Markdown and YAML can be done to some degree for example the `markdownlint-cli` tool can auto fix basic errors.  Whilst we are not a Python project there is a Python tool that does format YAML files.  From what I can see `Prettier` which is mainly written in JavaScript and can run on `pre-commit` is now best in class for code formatting. `Prettier` is an opinionated code formatter and has about 39,000+ Stars.  With `Prettier` there are five different ways to implement the `pre-commit` check.  Seems `Prettier` is the way to go at least for some `pre-commit` checks and it seems that one of the best features of `pre-commit` is the added automation.

What I am trying to say is sometimes in software development some of the best tools that you need are written in another language or not the main language that you use on a project. 

There are some projects for Lua which I have never used so not really sure about them but seems some don't support the correct versions of Lua.

https://github.com/igorshubovych/markdownlint-cli
https://pypi.org/project/yamlfmt/
https://prettier.io/
https://prettier.io/docs/en/precommit.html
https://gitlab.com/pablodiehl/pre-commit-lua-formatter
https://github.com/starwing/lua-fmt
https://luarocks.org/modules/xavier-wang/lua-fmt
https://luarocks.org/modules/luarocks/formatter

Screen Shot 2021-04-09 at 4 52 56 am

Screen Shot 2021-04-09 at 4 55 07 am

And from Prettier site:

Screen Shot 2021-04-09 at 5 18 47 am

Screen Shot 2021-04-09 at 5 06 04 am

Screen Shot 2021-04-09 at 5 06 16 am

Screen Shot 2021-04-09 at 5 19 15 am

**Comment by jbampton** [comment]
> The pre-commit framework is good.
> But does it add any value as we already check them in the CI? There are many duplicate rules between this framework and the CI files. I don't want to increase the maintain burden and make amateur harder to do contribution.

Hey @spacewander so seems maybe we are moving to use the GitHub Super-Linter on the CI since it provides more language checks or tests than the current GitHub Actions and is more robust and better supported.

So we should have more docs with all basic tools listed and advise people on the different ways they can contribute and test.

So we can have maybe three different levels of GitHub user or testing:

- amateur: for no local testing, running all tests with the GitHub Super Linter and not installing the `pre-commit` framework
- medium: some local testing with the `common` linting tools provided by both Shell and Batch files and running all tests with the GitHub Super Linter and not installing the `pre-commit` framework
- advanced: installing and using the `pre-commit` framework for local tests and for added automation with additional backup tests using the GitHub Super Linter for validation

> > It does say that `pre-commit` can be run on CI -> https://pre-commit.com/#usage-in-continuous-integration
> 
> Cool! I think it would be better if we could do that in this PR

So @Yiyiyimu is pushing for using  the `pre-commit` framework on the CI and maybe this is used instead of the GitHub Super Linter or perhaps together with it for even more tests and coverage.  Between the GitHub Super Linter,  `pre-commit` and  `Prettier` there seems to be a lot of things you do, check and test.

**Comment by jbampton** [comment]
Hey @membphis @spacewander @moonming @Yiyiyimu this is ready for review.

We have basic pre-commit framework which can now be used by advanced users and more tests can be added later or the tests can be made better or more relevant as we move along and all learn about pre-commit.

Who is going to join me in testing this and getting it merged ?

Install pre-commit and then test all files and checks with -> pre-commit run --all-files

Then start testing your commits with pre-commit.  

Some say seeing is believing so maybe once you start using pre-commit you will see even more benefits from it. 

I will finish the work on the GitHub Super Linter as well. 

We can also add Prettier later as well if needed.

**Comment by jbampton** [comment]
Hey @juzhiyuan please remove the `stale` label 👍 

Thank you.

**Comment by Yiyiyimu** [comment]
Hi @jbampton Thank you for your contribution and actually I quite like how pre-commit works and what it improves! Some points that I found could be better:
1. replace the original Actions with pre-commit in CI in this PR. Since we don't need two tools to do the same thing, it's better to do the replacement rather than just addition. It's also the same about Super-Linter.
2. add a script and make it could be called from Makefile to install pre-commit locally, so users would not be confusing about how to install it.
3. There are quite a lot of checks for python files, AFAIK `debug-statements` / `https://github.com/psf/black` / `https://github.com/asottile/dead` which I don't think is necessary for this repo

**Comment by jbampton** [comment]
Hey @Yiyiyimu can you please review ?

I have removed all the Python checks from the pre-commit and I have removed the original GitHub Actions for markdownlint and yamlint.  I added pre-commit to the CI.

I added a new check to pre-commit called codespell which found some more spelling mistakes and we now have a small custom dictionary.  I cleaned up the lint file removing the emoji since I think it might look better.

pre-commit has a check for trailing whitespace but it doesn not give us as much coverage as our custom GitHub action since with pre-commit we ignore the file but with actions we ignore lines in that file.  Still good to at least have a check on pre-commit for trailing whitespace.  You can even write your own hook scripts and on Apache Airflow they write them in Shell and Python. 

Pre-commit can be installed with `pip`, `curl`, `brew` or `conda` so maybe we should leave it up to the user to install since they might prefer a certain way.    I am willing to add some more docs too.

File: codespell.txt
Comment (spacewander) [inline codespell.txt:1]: Is this file necessary? It would be better to skip `docs/es` instead of putting Spanish words in the ignore list.

Comment (jbampton) [reply codespell.txt:1]: Yes the file is necessary but you are correct that we don't need Spanish words in the ignore list.

We use all lowercase "words" in the ignore list.  I just now ran:

```
codespell --skip="./docs/es,./docs/zh,./.git,./**/*.crt,./**/go.sum" | cut -d' ' -f2 | sort | uniq        
WithT
hel
nD
nd
nin
witht
```

File: .pre-commit-config.yaml
Comment (spacewander) [inline .pre-commit-config.yaml:52]: It would be easier to maintain if you use an allow list. There are many things in the `.gitignore` that need to be excluded.

File: CONTRIBUTING.md
Comment (spacewander) [inline CONTRIBUTING.md:89]: Could we remove this so that we don't need to maintain two versions of codespell? People can know how to run it by reading the CI setup.

File: .pre-commit-config.yaml
Comment (jbampton) [reply .pre-commit-config.yaml:52]: Can you give me a list of files and/or folders I should check or allow ?

File: .gitignore
Comment (Yiyiyimu) [inline .gitignore:68]: might be good to keep the extra newline

File: .pre-commit-config.yaml
Comment (Yiyiyimu) [inline .pre-commit-config.yaml:51]: I'm a bit confused about this place. If pre-commit finds a place that need to add a license, would it add with bot account of Github Action? If so, the action is prohibited by Apache Community since we do not allow bot to do commits.

File: .pre-commit-config.yaml
Comment (Yiyiyimu) [inline .pre-commit-config.yaml:72]: I can't get it what does this `Other` means here, since there are still quite a lot categories after this one

File: CONTRIBUTING.md
Comment (Yiyiyimu) [inline CONTRIBUTING.md:59]: Suggest removing this line since it's not related to contributing guide

File: CONTRIBUTING.md
Comment (Yiyiyimu) [inline CONTRIBUTING.md:57]: I still believe it's better to add a universal script for contributors to easily use it, like using `curl` to do the installation.

File: codespell.txt
Comment (Yiyiyimu) [inline codespell.txt:1]: Sorry I failed to understand this. Is this a whitelist for misspell?

**Review by Yiyiyimu — COMMENTED** [review]
plz do not force push at the middle of reviewing, or I have totally no idea about what's got changed

File: codespell.txt
Comment (jbampton) [reply codespell.txt:1]: `codespell` is a new Python based spell checker that runs on `pre-commit`.  This is the `codespell` dictionary and it found some new spelling mistakes.  I added docs on this

File: .pre-commit-config.yaml
Comment (spacewander) [reply .pre-commit-config.yaml:51]: @jbampton 
Please remove the license checker. We already have one. There is no need to add an extra template file. And your PR is incorrect. Please keep a **fixed and limit** scope to your PR. **I can't review a PR that does not have a fixed goal and change rapidly.**

File: apisix/balancer/ewma.lua
Comment (spacewander) [inline apisix/balancer/ewma.lua:1]: This is wrong. This file's LICENSE has been specified in the LICENSE separately.

File: .pre-commit-config.yaml
Comment (spacewander) [inline .pre-commit-config.yaml:161]: Why exclude?

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
@jbampton 
If you want your PR to be accepted, please limit its scope. If you keep adding new things with `force push` which destroys the history, I can't trust you and leave approval (as you may add new things just after I have reviewed it).

**Comment by jbampton** [comment]
This is not a license checker.  This code automates adding the license on `git commit` if it is missing and therefore the existing license checker on build should never fail.  `pre-commit` is all about automation.

File: .pre-commit-config.yaml
Comment (jbampton) [reply .pre-commit-config.yaml:161]: `luacheck` seems to fail on `pre-commit` if you include these files

File: .pre-commit-config.yaml
Comment (jbampton) [reply .pre-commit-config.yaml:72]: We are going to group all "Other" file types that are common and have the same type of license.

So I just grouped some minor file types to start with.

File: .pre-commit-config.yaml
Comment (jbampton) [reply .pre-commit-config.yaml:51]: We are not using bots.   I am just using it exactly like Apache Airflow and it's allowed.  

When run `pre-commit` on the CI the license check would just fail if there was a missing license and the `pre-commit` GitHub Action would fail.   When run locally if the license check fails then `pre-commit` would auto insert the missing license.

File: CONTRIBUTING.md
Comment (jbampton) [reply CONTRIBUTING.md:57]: Most people would not install `pre-commit` with `curl` if they had the choice to use `pip`, `brew` or `conda`.  I use Homebrew and I update my packages everyday and it makes it so much easier to use and manage.

File: CONTRIBUTING.md
Comment (jbampton) [reply CONTRIBUTING.md:59]: Having as much information as possible on `pre-commit` is necessary.   
So far both you and @spacewander don't seem to understand much about `pre-commit`. 
Perhaps we need a new Markdown file for these docs ?

**Comment by Yiyiyimu** [comment]
For real? Force push again?

**Pull request closed without merging**
